src/Optimization/optimization_problem_results.jl

```diff
@@ -305,55 +305,87 @@ function _copy_for_serialization(res::OptimizationProblemResults)
 end
 
 function _read_results(
-    result_values::Dict{<:OptimizationContainerKey, DataFrames.DataFrame},
+    result_values::Dict{<:OptimizationContainerKey, DataFrame},
     container_keys,
     timestamps::Vector{Dates.DateTime},
     time_ids,
     base_power::Float64,
     base_timestamps::Vector{Dates.DateTime},
+    table_format::TableFormat,
 )
     existing_keys = keys(result_values)
     container_keys = container_keys === nothing ? existing_keys : container_keys
     _validate_keys(existing_keys, container_keys)
-    results = Dict{OptimizationContainerKey, DataFrames.DataFrame}()
-    for (k, v) in result_values
-        if k in container_keys
-            num_rows = DataFrames.nrow(v)
-            if num_rows == 1 && num_rows < length(time_ids)
-                results[k] =
-                    if convert_result_to_natural_units(k)
-                        v .* base_power
-                    else
-                        v
-                    end
-            else
-                total_rows = size(v)[1]
-                if length(base_timestamps) == total_rows
-                    results[k] =
-                        if convert_result_to_natural_units(k)
-                            v[time_ids, :] .* base_power
-                        else
-                            v[time_ids, :]
-                        end
-                    DataFrames.insertcols!(results[k], 1, :DateTime => timestamps)
-                else
-                    @warn(
-                        "Length of variables is different than timestamps. Ignoring timestamps."
+    results = Dict{OptimizationContainerKey, DataFrame}()
+    IS.@assert_op time_ids == collect(1:length(timestamps))
+    df_timestamps = DataFrame(:DateTime => timestamps, :time_index => time_ids)
+
+    for (key, df) in result_values
+        if !in(key, container_keys)
+            continue
+        end
+        component_cols = [:component]
+        if "component_x" in names(df)  # TODO DT: what should this name be?
+            push!(component_cols, "component_x")
+            if table_format == TableFormat.WIDE
+                error(
+                    "Wide format is not supported with 3-dimensional results",
+                )
+            end
+        end
+        num_components = DataFrames.nrow(unique(df[:, component_cols]))
+        num_rows = DataFrames.nrow(df)
+        if num_rows % num_components != 0
+            error(
+                "num_rows = $num_rows is not divisible by num_components = $num_components",
+            )
+        end
+        num_rows_per_component = num_rows ÷ num_components
+        if num_rows_per_component == 1 && num_rows_per_component < length(time_ids)
+            # Case 1: Can't add a time column. TODO: Explain why this happens.
```

Comment:
The number of rows is less than the number of timestamps. Can happen when one variable in the model has 5-minute resolution and another variable has 1-hour resolution.